pkg/sfsturbo/services/volumes.go

```diff
@@ -0,0 +1,226 @@
+package services
+
+import (
+	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/config"
+	"github.com/kubernetes-csi/csi-lib-utils/protosanitizer"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"strconv"
+	"time"
+
+	"github.com/chnsz/golangsdk"
+	"github.com/chnsz/golangsdk/openstack/sfs_turbo/v1/shares"
+	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/common"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	log "k8s.io/klog/v2"
+)
+
+const (
+	shareCreating  = "100"
+	shareAvailable = "200"
+
+	shareSubExpanding     = "121"
+	shareSubExpandSuccess = "221"
+	shareSubExpandError   = "321"
+
+	shareDescription = "provisioned-by=sfsturbo.csi.huaweicloud.org"
+)
+
+func CreateShareCompleted(c *config.CloudCredentials, createOpts *shares.CreateOpts) (
+	*shares.TurboResponse, error) {
+	turboResponse, err := CreateShare(c, createOpts)
+	if err != nil {
+		return nil, err
+	}
+	log.V(4).Infof("[DEBUG] create share response detail: %v", protosanitizer.StripSecrets(turboResponse))
+	err = WaitForShareAvailable(c, turboResponse.ID)
+	if err != nil {
+		return nil, err
+	}
+	return turboResponse, nil
+}
+
+const (
+	DefaultInitDelay = 15 * time.Second
+	DefaultFactor    = 1.02
+	DefaultSteps     = 30
+)
+
+// WaitForShareAvailable will cost few minutes
+func WaitForShareAvailable(c *config.CloudCredentials, shareID string) error {
+	condition := func() (bool, error) {
+		share, err := GetShare(c, shareID)
+		if err != nil {
+			return false, status.Errorf(codes.Internal,
+				"Failed to query share %s when wait share available: %v", shareID, err)
+		}
+		log.V(4).Infof("[DEBUG] WaitForShareAvailable query detail: %v", protosanitizer.StripSecrets(share))
+		if share.Status == shareAvailable {
+			return true, nil
+		}
+		if share.Status == shareCreating {
+			return false, nil
+		}
+		return false, status.Errorf(codes.Internal, "created share status is not available : %s", share.Status)
+	}
+	backoff := wait.Backoff{
+		Duration: DefaultInitDelay,
+		Factor:   DefaultFactor,
+		Steps:    DefaultSteps,
+	}
+	return wait.ExponentialBackoff(backoff, condition)
+}
+
+func CreateShare(c *config.CloudCredentials, createOpts *shares.CreateOpts) (*shares.TurboResponse, error) {
+	createOpts.Description = shareDescription
+	client, err := getSFSTurboV1Client(c)
+	if err != nil {
+		return nil, err
+	}
+	share, err := shares.Create(client, createOpts).Extract()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to create share, err: %v", err)
+	}
+	return share, nil
+}
+
+func DeleteShareCompleted(c *config.CloudCredentials, shareID string) error {
+	if _, err := GetShare(c, shareID); err != nil {
```

Comment:
there should be no need to query, the API will return 404 when the resource does not exist

Reply:
For DeleteShare, 400 will be returned when the resource does not exist.
Under the same conditions, GetShare will return 404